File: android/Detonator/src/main/java/com/iconshot/detonator/RequestPromise.java
package com.iconshot.detonator;

import java.util.ArrayList;
import java.util.List;

public class RequestPromise {
    private Detonator detonator;
    private int fetchId;

    public RequestPromise(Detonator detonator, int fetchId) {
        this.detonator = detonator;
        this.fetchId = fetchId;
    }

    public void resolve() {
        resolve("");
    }

    public void resolve(Object data) {
        List<Object> lines = new ArrayList();

        lines.add(String.valueOf(fetchId));
        lines.add(data);

        String value = detonator.messageFormatter.join(lines);

        detonator.uiHandler.post(() -> {
            detonator.send("com.iconshot.detonator.request.fetch::resolve", value);
        });
    }

    public void reject(Exception exception) {
        reject(exception.getMessage());
    }

    public void reject(String message) {
        List<Object> lines = new ArrayList();

        lines.add(String.valueOf(fetchId));
        lines.add(message);

        String value = detonator.messageFormatter.join(lines);

        detonator.uiHandler.post(() -> {
            detonator.send("com.iconshot.detonator.request.fetch::reject", value);
        });
    }
}

File: ios/Detonator/Sources/Detonator/RequestPromise.swift
import Foundation

public class RequestPromise {
    private let detonator: Detonator
    private let fetchId: Int
    
    init(_ detonator: Detonator, _ fetchId: Int) {
        self.detonator = detonator
        self.fetchId = fetchId
    }
    
    public func resolve(_ data: Encodable? = "") -> Void {
        let lines: [Encodable?] = [fetchId, data]
        
        let value = detonator.messageFormatter.join(lines)
        
        DispatchQueue.main.async {
            self.detonator.send("com.iconshot.detonator.request.fetch::resolve", value)
        }
    }
    
    public func reject(_ error: Error) -> Void {
        reject(error.localizedDescription)
    }
    
    public func reject(_ message: String) -> Void {
        let lines: [Encodable?] = [fetchId, message]
        
        let value = detonator.messageFormatter.join(lines)
        
        DispatchQueue.main.async {
            self.detonator.send("com.iconshot.detonator.request.fetch::reject", value)
        }
    }
}
